keps/sig-node/4939-tls-in-grpc-probe/README.md

```diff
@@ -0,0 +1,682 @@
+-->
+
+The primary goal is to support TLS connections when using the `grpc` probe.
+The probe will use TLS but not verify the certificate.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+It is not a goal of this KEP to support providing a certificate to verify the TLS
+connection.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+I would like to add new configuration fields alongside `port` and `service` in the
+[Probe GRPCAction](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#Probe).
+
+They can be used to indicate whether or not TLS should be used
+and can serve as a basis for future TLS related functionality if desired.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+1. Adds more code to Kubelet and surface area to `Pod.Spec`
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We should add a new struct to
+[GRPCAction](https://github.com/kubernetes/kubernetes/blob/f422a58398d457a88ac7b05afca85a0e6b805605/pkg/apis/core/types.go#L2665-L2676)
+named `tls` which is optional.
+It would have a single field `mode` which only will have one value currently `NoVerify`.
+
+The presence of the `tls` object would be enough to indicate a desire to use TLS.
+
+For example, these configurations would enable TLS with no verification:
+
+```
+grpcAction:
+    port: 12345
+    tls:
+        mode: NoVerify
+```
+```
+grpcAction:
+    port: 12345
+    tls:
+```
+
+This configuration would disable TLS:
+
+```
+grpcAction:
+    port: 12345
+```
+
+I have identified the relevant code changes once the configuration
+structure is hammered out.
+
+Currently, the gRPC probe uses `insecure.NewCredentials()` when
+[establishing the connection](https://github.com/kubernetes/kubernetes/blob/d9441212d3e11dc13198f9d4df273c3555ecad11/pkg/probe/grpc/grpc.go#L59).
+
+If configured to use TLS, that `DialOption` should be replaced with:
+
+```go
+tlsConfig := &tls.Config{
+    InsecureSkipVerify: true,
```

Comment:
So the plan to only add this if the connection requires TLS?

I tried this before: https://github.com/kubernetes/kubernetes/pull/124522/

Also gRPC suggested that the option to ignore cert errors  that may work for both - tls and non tls can be implemented: https://groups.google.com/g/grpc-io/c/dsW1AOyzZX0 Perhaps we can just ignore all errors like this?

Reply:
ok, I hoped we can get away with less config.